lib/streamlit/runtime/caching/storage/local_disk_cache_storage.py

```diff
@@ -0,0 +1,174 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import math
+import os
+import shutil
+
+from streamlit import util
+from streamlit.file_util import get_streamlit_file_path, streamlit_read, streamlit_write
+from streamlit.logger import get_logger
+from streamlit.runtime.caching.storage.cache_storage_protocol import (
+    CacheStorage,
+    CacheStorageContext,
+    CacheStorageError,
+    CacheStorageKeyNotFoundError,
+    CacheStorageManager,
+)
+from streamlit.runtime.caching.storage.in_memory_cache_storage_wrapper import (
+    InMemoryCacheStorageWrapper,
+)
+
+# Streamlit directory where persisted @st.cache_data objects live.
+# (This is the same directory that @st.cache persisted objects live.
+# But @st.cache_data uses a different extension, so they don't overlap.)
+_CACHE_DIR_NAME = "cache"
+
+# The extension for our persisted @st.cache_data objects.
+# (`@st.cache_data` was originally called `@st.memo`)
+_CACHED_FILE_EXTENSION = "memo"
+
+_LOGGER = get_logger(__name__)
+
+
+class LocalDiskCacheStorageManager(CacheStorageManager):
+    def create(self, context: CacheStorageContext) -> CacheStorage:
+        """Creates a new cache storage instance wrapped with in-memory cache facade"""
+        persist_storage = LocalDiskCacheStorage(context)
+        return InMemoryCacheStorageWrapper(
+            persist_storage=persist_storage, context=context
+        )
+
+    def clear_all(self) -> None:
+        cache_path = get_cache_folder_path()
+        if os.path.isdir(cache_path):
+            shutil.rmtree(cache_path)
+
+    def check_context(self, context: CacheStorageContext) -> None:
+        if (
+            context.persist == "disk"
+            and context.ttl_seconds is not None
+            and not math.isinf(context.ttl_seconds)
+        ):
+            _LOGGER.warning(
+                f"The cached function '{context.function_display_name}' has a TTL "
+                "that will be ignored. Persistent cached functions currently don't "
+                "support TTL."
+            )
+
+
+class LocalDiskCacheStorage(CacheStorage):
+    """
+    Cache storage that persists data to disk
+    This is the default cache storage for @st.cache_data
```

Comment:
@tconkling, @vdonato I would like to explain the original idea:

Storage should not know about any details of its use, and only provide the ability to add and remove entries.

In this case, the `LocalDiskCacheStorage` class purely provides the ability to work with disk storage. 
It is only the responsibility of LocalDiskCacheStorage class. 

In contrary `LocalDiskCacheStorageManager` is responsible for creating those storages, and in our case, it is wrapping those storage instances with another layer of in-memory cache via `InMemoryCacheWrapper`. `LocalDiskCacheStorageManager` was originally called `InMemoryWrappedLocalDiskCacheStorageManager` :) but then during a review, we renamed that to `LocalDiskCacheStorage` for simplicity). 

So the idea is that each implementation should consist of `ConcreteCacheStorage`, and `ConcreteCacheStorageManager` , responsible for the creation (and optional decoration) of ConcreteCacheStorage.  

For convenience, we provide `InMemoryCacheStorageWrapper` class, which could wrap any storage with an in-memory cache layer, but this is a detail of the manager, not the storage itself.

I think this is my flaw in the description of the classes, it will be necessary to work on this improvement